src/display/editor/tools.js

```diff
@@ -78,75 +84,60 @@ class CommandManager {
     overwriteIfSameType = false,
     keepUndo = false,
   }) {
+    if (mustExec) {
+      cmd();
+    }
+
     const save = { cmd, undo, type };
-    if (
-      overwriteIfSameType &&
-      !isNaN(this.#position) &&
-      this.#commands[this.#position].type === type
-    ) {
+    if (this.#position === -1) {
+      this.#position = 0;
+      this.#commands.push(save);
+      return;
+    }
+
+    if (overwriteIfSameType && this.#commands[this.#position].type === type) {
       // For example when we change a color we don't want to
       // be able to undo all the steps, hence we only want to
       // keep the last undoable action in this sequence of actions.
       if (keepUndo) {
         save.undo = this.#commands[this.#position].undo;
       }
       this.#commands[this.#position] = save;
-      if (mustExec) {
-        cmd();
-      }
       return;
     }
-    const next = (this.#position + 1) % this.#maxSize;
-    if (next !== this.#start) {
-      if (this.#start < next) {
-        this.#commands = this.#commands.slice(this.#start, next);
-      } else {
-        this.#commands = this.#commands
-          .slice(this.#start)
-          .concat(this.#commands.slice(0, next));
+
+    const next = this.#position + 1;
+    if (next === this.#maxSize) {
+      this.#commands.splice(0, 1);
+    } else {
+      this.#position = next;
+      if (next < this.#commands.length) {
+        this.#commands.splice(next);
```

Comment:
*I might be missing obvious something, but:*
Above you're removing from the *start* of the `commands`-array, but here it seems that you instead remove from its *end*. Is that actually intended, since I'd sort of expect that you'd always want to remove the oldest commands first?

Reply:
It should be correct like this.
For example:
 - add A (commands = [A], position = 0)
 - add B (commands = [A, B], position = 1)
 - add C (commands = [A, B, C], position = 2)
 - undo (commands = [A, B, C], position = 1)
 - undo (commands = [A, B, C], position = 0)
 - add D (commands = [A, D], position = 1)

Before adding D, we remove B and C from (0+1) to the end.
Normally undo/redo works like this, but of course I can be wrong.

Reply:
I'm going to blame the current heatwave, since I've been unable to think straight most of the day :-)